File: Backtracking/46_Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        backtract(list, new ArrayList<Integer>(), nums);
        return list;
    }

    public void backtract(List<List<Integer>> list, List<Integer> tmplist, int[] nums) {
        if (tmplist.size() == nums.length) {
            list.add(new ArrayList<Integer>(tmplist));
        } else {
            for (int i=0; i<nums.length; i++) {
                if (tmplist.contains(nums[i])) {
                    continue;
                }
                tmplist.add(nums[i]);
                backtract(list, tmplist, nums);
                tmplist.remove(tmplist.size() - 1);
            }
        }
    }
}

File: Backtracking/46_Permutations.py
import copy
class Solution(object):

    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        lst, tmplst = [], []
        self.backtrack(lst, tmplst, nums)
        return lst

    def backtrack(self, lst, tmplst, nums):
        if len(tmplst) == len(nums):
            # must use deep copy
            lst.append(copy.deepcopy(tmplst))
        else:
            for num in nums:
                if num in tmplst:
                    continue
                tmplst.append(num)
                self.backtrack(lst, tmplst, nums)
                tmplst.pop()
